lib/ansible/modules/source_control/gitlab_project_variable.py

```diff
@@ -0,0 +1,192 @@
+version_added: "2.9"
+author:
+    - "Markus Bergholz (@markuman)"
+requirements:
+  - python >= 2.7
+  - python-gitlab python module
+options:
+  state:
+    description:
+      - create or delete project variable.
+      - Possible values are present and absent.
+    default: present
+    type: str
+    choices: ["present", "absent"]
+  server_url:
+    description:
+      - The URL of the Gitlab server, with protocol (i.e. http or https).
+    required: true
+    type: str
+  login_token:
+    description:
+      - Gitlab access token with api permissions.
+    required: true
+    type: str
+  name:
+    description:
+      - The path and name of the project
+    required: true
+    type: str
+  purge_vars:
+    description:
+      - When set to true, all variables which are not presented in the task will be deleted.
+    default: false
+    required: false
+    type: bool
+  vars:
+    description:
+      - A list of key value pairs
+    default: []
+    required: false
+    type: list
+'''
+
+
+EXAMPLES = '''
+- name: Set or update some CI/CD variables
+  gitlab_project_variables:
+    server_url: https://gitlab.com
+    login_token: secret_access_token
+    name: markuman/dotfiles
+    purge_vars: False
+    vars:
+      - ACCESS_KEY_ID: abc123
+      - SECRET_ACCESS_KEY: 321cba
+
+- name: delete one variable
+  gitlab_project_variables:
+    server_url: https://gitlab.com
+    login_token: secret_access_token
+    name: markuman/dotfiles
+    state: absent
+    vars:
+      - ACCESS_KEY_ID: abc123
+'''
+
+RETURN = '''# '''
+
+from ansible.module_utils.basic import AnsibleModule, missing_required_lib
+
+try:
+    import gitlab
+    HAS_GITLAB_PACKAGE = True
+except Exception:
+    HAS_GITLAB_PACKAGE = False
+
+
+class gitlab_project_variables(object):
+
+    def __init__(self, login_token, project_name, server_url):
+        self.repo = gitlab.Gitlab(
+            '{server_url}'.format(server_url=server_url), private_token=login_token)
+        self.repo.auth()
+        self.project = self.get_project(project_name)
+
+    def get_project(self, project_name):
+        return self.repo.projects.get(project_name)
+
+    def list_all_project_variables(self):
+        raw_variable_list = self.project.variables.list()
+        retval = []
+        if len(raw_variable_list) > 0:
+            for item in raw_variable_list:
+                retval.append(item.get_id())
+        return retval
+
+    def create_variable(self, key, value):
+        return self.project.variables.create({"key": key, "value": value})
+
+    def update_variable(self, key, value):
+        var = self.project.variables.get(key)
+        if var.value == value:
+            return False
+        var.save()
+        return True
+
+    def delete_variable(self, key):
+        return self.project.variables.delete(key)
+
+
+def native_python_main(server_url, login_token, project_name, purge_vars, var_list, state):
+
+    change = False
+    this_gitlab = gitlab_project_variables(
```

Comment:
Put the object creation out of this function in the `main` to be more readable.
Then only pass the gitlab object to your function.

Reply:
https://github.com/markuman/ansible/commit/d861dad27d61d89f8c992355f532d97dbcbc6188  
https://github.com/markuman/ansible/commit/ae0157ddc21bab5120c1e71a0cb8c50c686c7c82